### PR TITLE
[AutoDiff] IRGen differentiability witness tables

### PR DESCRIPTION
**Pull request by marcrasi**
Summary of changes:
* Add `LinkEntity` for `SILDifferentiabilityWitnessTable`.
* Add `IRGenModule::emitSILDifferentiabilityWitness` & `IRGenModule::getAddrOfDifferentiabilityWitness`, which do the actual work of emitting the IR.
* Add TBDGen for the differentiability witnesses. Also some factoring duplicated code out into a common function.

I think this resolves TF-930, by making TBDGen incorporate the VJP/JVP generic signatures in its calculation.

### CODE REVIEW

**Review by dan-zheng — APPROVED** [review]
Nice TBDGen refactoring!
Closing TF-930 sounds good if it's no longer reproducible.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please clean test tensorflow

**Comment by marcrasi** [comment]
The build error looks related to the monorepo transition. I expect that rebasing this on top of the transition will fix it. I'll wait until the transition gets merged into tensorflow (it should be really soon, right?), rebase this PR, and then rerun CI on it.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please clean test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Review by rxwei — APPROVED** [review]
LGTM. Thanks for keeping me in the loop!

File: lib/AST/PrettyStackTrace.cpp
Comment (dan-zheng) [inline lib/AST/PrettyStackTrace.cpp]:
```suggestion
    bool addNewline) {
```
`ASTContext &Context` is unused, please remove the argument here and the corresponding field from `PrettyStackTraceDifferentiabilityWitness`.

Comment (marcrasi) [reply lib/AST/PrettyStackTrace.cpp]: done

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow linux

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Pull request merged**
